### ISSUE TITLE
[SQLancer] Wrong result with Float->Int cast in WHERE

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#18410]
### What version of Materialize are you using?

v0.48.0-dev (f7027928e)

### How did you install Materialize?

Built from source

### What is the issue?

Wrong result found by SQLancer:
```
--java.lang.AssertionError: the results mismatch!
---- SELECT SUM(0.9910018062404208905036284704692661762237548828125) FROM t2;
---- result: 10.9010198686446297955399131751619279385
---- SELECT SUM(agg0) FROM (SELECT SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE (CAST(t2.c0 AS INT) != 0) UNION ALL SELECT SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE NOT (((t2.c0)::INT != 0)) UNION ALL SELECT ALL SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE ((CAST(t2.c0 AS INT) != 0)) ISNULL) as asdf;
---- result: 7.92801444992336712402902776375412940979
--	at sqlancer.postgres.oracle.tlp.PostgresTLPAggregateOracle.aggregateCheck(PostgresTLPAggregateOracle.java:85)
--	at sqlancer.postgres.oracle.tlp.PostgresTLPAggregateOracle.check(PostgresTLPAggregateOracle.java:48)
--	at sqlancer.common.oracle.CompositeTestOracle.check(CompositeTestOracle.java:22)
--	at sqlancer.ProviderAdapter.generateAndTestDatabaseWithQueryPlanGuidance(ProviderAdapter.java:132)
--	at sqlancer.Main$DBMSExecutor.run(Main.java:362)
--	at sqlancer.Main$2.run(Main.java:559)
--	at sqlancer.Main$2.runThread(Main.java:541)
--	at sqlancer.Main$2.run(Main.java:532)
--	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
--	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
--	at java.base/java.lang.Thread.run(Thread.java:829)
---- Time: 2023/03/25 06:30:42
-- Database: database1
-- Database version: 9.5.0
-- seed value: 1679698092783
\c test;
DROP DATABASE IF EXISTS wrong;
CREATE DATABASE wrong;
\c wrong;
CREATE TABLE t0(c0 VARCHAR(223) );
CREATE TABLE IF NOT EXISTS t1(c0 FLOAT );
CREATE TABLE IF NOT EXISTS t2(c0 FLOAT );
CREATE TABLE t3(c0 INT     );
DELETE FROM t3;
CREATE INDEX t3i0 ON t3(c0);
INSERT INTO t1(c0) VALUES(0.6129513), (0.14332032), (0.7458146);
INSERT INTO t3(c0) VALUES(((-1605372624)+(-1793888592)));
INSERT INTO t1(c0) VALUES(0.44028056);
DROP INDEX IF EXISTS t0i0 RESTRICT;
DROP INDEX IF EXISTS t1i2 RESTRICT;
CREATE INDEX t3i1 ON t3(c0 DESC, c0);
INSERT INTO t2(c0) VALUES(0.5487458);
INSERT INTO t2(c0) VALUES(- (CAST(((((362209324)>>(-203092041)))&(((1004145210)<<(535604325)))) AS INT))), (0.6520862), (1.20671219E9);
INSERT INTO t2(c0) VALUES(-2.03092048E8), (0.19686678), (0.96729666);
DROP INDEX IF EXISTS t0i0 RESTRICT;
INSERT INTO t2(c0) VALUES(0.49358597), (1.20671219E9), (-1.79915456E8);
INSERT INTO t1(c0) VALUES(0.5487458);
INSERT INTO t0(c0) VALUES('C|6\');
INSERT INTO t3(c0) VALUES(362209324);
UPDATE t0 SET c0=(t0.c0) WHERE ((t0.c0)=(t0.c0));
INSERT INTO t2(c0) VALUES(0.024532218);
CREATE INDEX t2i0 ON t2(c0 ASC NULLS LAST);
UPDATE t0 SET c0='' WHERE (t0.c0) BETWEEN (t0.c0) AND (t0.c0);
SELECT SUM(0.9910018062404208905036284704692661762237548828125) FROM t2;-- result: 10.9010198686446297955399131751619279385
SELECT SUM(agg0) FROM (SELECT SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE (CAST(t2.c0 AS INT) != 0) UNION ALL SELECT SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE NOT (((t2.c0)::INT != 0)) UNION ALL SELECT ALL SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE ((CAST(t2.c0 AS INT) != 0)) ISNULL) as asdf;-- result: 7.92801444992336712402902776375412940979;
```
This is different from https://github.com/MaterializeInc/materialize/issues/18346 since all queries are successful here.
The last part of the `UNION` seems to be wrong, when I run it in isolation:
```
wrong=> SELECT ALL SUM(0.9910018062404208905036284704692661762237548828125)  as agg0 FROM t2 WHERE ((CAST(t2.c0 AS INT) != 0)) IS NULL;
 agg0
------

(1 row)
```
There is 1 row shown as the result, but no actual result returned. So next I checked the values in `t2` and the parts of the subqueries:
```
wrong=> select * from t2;
     c0
-------------
           0
  -179915456
  -203092048
  1206712190
  1206712190
  0.49358597
  0.96729666
   0.6520862
   0.5487458
  0.19686678
 0.024532218
(11 rows)

wrong=> select * from t2 where (CAST(t2.c0 AS INT) != 0);
     c0
------------
 -179915456
 -203092048
 1206712190
 1206712190
 0.96729666
  0.6520862
  0.5487458
(7 rows)

wrong=> select * from t2 where NOT (((t2.c0)::INT != 0));
 c0
----
  0
(1 row)

wrong=> select * from t2 where ((CAST(t2.c0 AS INT) != 0)) IS NULL;
 c0
----
(0 rows)
```
This shows another problem with the `select * from t2 where NOT (((t2.c0)::INT != 0));` query. It should contain 4 values, see:
```
wrong=> select *, not (cast(t2.c0 as int) != 0) from t2;
     c0      | ?column?
-------------+----------
           0 | t
  -179915456 | f
  -203092048 | f
  1206712190 | f
  1206712190 | f
  0.49358597 | t
  0.96729666 | f
   0.6520862 | f
   0.5487458 | f
  0.19686678 | t
 0.024532218 | t
(11 rows)
```
But in truth only contains one:
```
wrong=> select * from t2 where NOT (((t2.c0)::INT != 0));
 c0
----
  0
(1 row)
```
My guess is that the `0.xx` values are wrongly converted to Int when inside the `where` and are thus compared as if they still are floats, not integers.

**Comment by parker-timmerman** [comment]
I looked into this issue, and it seems like it's a bug in our optimizer. For example, if we initialize our state with:
```
CREATE TABLE t1 ( c0 FLOAT ); INSERT INTO t1 VALUES (-0.1), (0.6), (1), (0);
```
and then run the problematic query...
```
SELECT * FROM t1 WHERE NOT (((t1.c0)::INT != 0));
```
...we see the correct results!
```
  c0
------
    0
 -0.1
(2 rows)
```

But! Once we add an index...
```
CREATE INDEX t1i0 ON t1 (c0);
```

🚨 ... our problematic query starts to fail 🚨
```
materialize=> SELECT * FROM t1 WHERE NOT (((t1.c0)::INT != 0));
 c0
----
  0
(1 row)
```

If we drop the `NOT` from the query, then we get back the right results, for this query:
```
materialize=> SELECT * FROM t1 WHERE (((t1.c0)::INT != 0));
 c0
-----
   1
 0.6
(2 rows)
```
---
I do not know how the optimizer works, but some rudimentary debugging I found the following:
1. The issue seems to be in the [physical_optimimzer](https://github.com/MaterializeInc/materialize/blob/main/src/transform/src/dataflow.rs#L60)
2. After the `physical_optimizer` runs, what used to be a `MirRelationExpr::Filter { ... }` gets transformed to `MirRelationExpr::Project {  input: Join { ... }, implementation: IndexedFilter { ... }, ... }`, **but only when** the predicate for the original `Filter { ... }` is `BinaryFunc::Eq`. This transform doesn't seem to occur when the predicate is `BinaryFunc::NotEq`, which I'm guessing is why the final query from above, works.
3. Commenting out the [LiteralConstraints](https://github.com/MaterializeInc/materialize/blob/main/src/transform/src/lib.rs#L546) transform fixes the issue. 
4. So my _guess_ is it's something wrong with the `MapFilterProject` optimization, but I could be missing something here

For now I'm going to assign this issue back to you @antiguru since it seems like it's not a type system issue? More than happy to take it back though if it is 🙂
